File: lib/core/services/analytics_service.dart
import 'package:flutter/foundation.dart';
import 'package:mixpanel_flutter/mixpanel_flutter.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Analytics service for tracking user events using Mixpanel
/// Respects GDPR opt-out preferences
class AnalyticsService {
  static Mixpanel? _mixpanel;
  static bool _isInitialized = false;
  static bool _analyticsEnabled = true;

  /// Initialize Mixpanel SDK
  static Future<void> initialize({required String token}) async {
    if (_isInitialized) return;

    try {
      _mixpanel = await Mixpanel.init(
        token,
        trackAutomaticEvents: false, // Manual tracking for GDPR compliance
      );
      _isInitialized = true;

      // Check GDPR consent
      await _loadAnalyticsConsent();

      debugPrint('Analytics service initialized');
    } catch (e) {
      debugPrint('Failed to initialize analytics: $e');
    }
  }

  /// Load analytics consent from local storage
  static Future<void> _loadAnalyticsConsent() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final analyticsConsent = prefs.getBool('gdpr_analytics_consent');
      
      // If no explicit consent is given, default to disabled (opt-in approach)
      _analyticsEnabled = analyticsConsent ?? false;

      if (!_analyticsEnabled) {
        await optOut();
      }
    } catch (e) {
      debugPrint('Error loading analytics consent: $e');
      _analyticsEnabled = false;
    }
  }

  /// Update analytics consent status
  static Future<void> updateConsent(bool enabled) async {
    _analyticsEnabled = enabled;
    
    if (!enabled) {
      await optOut();
    } else {
      await optIn();
    }
  }

  /// Opt out of analytics tracking (GDPR compliance)
  static Future<void> optOut() async {
    if (_mixpanel == null) return;
    
    try {
      _mixpanel!.optOutTracking();
      debugPrint('Analytics tracking opted out');
    } catch (e) {
      debugPrint('Error opting out of analytics: $e');
    }
  }

  /// Opt in to analytics tracking
  static Future<void> optIn() async {
    if (_mixpanel == null) return;
    
    try {
      _mixpanel!.optInTracking();
      debugPrint('Analytics tracking opted in');
    } catch (e) {
      debugPrint('Error opting in to analytics: $e');
    }
  }

  /// Identify a user
  static Future<void> identify(String userId) async {
    if (!_isInitialized || !_analyticsEnabled || _mixpanel == null) return;

    try {
      await _mixpanel!.identify(userId);
    } catch (e) {
      debugPrint('Error identifying user: $e');
    }
  }

  /// Set user properties
  static Future<void> setUserProperties(Map<String, dynamic> properties) async {
    if (!_isInitialized || !_analyticsEnabled || _mixpanel == null) return;

    try {
      final people = _mixpanel!.getPeople();
      await people.set(properties);
    } catch (e) {
      debugPrint('Error setting user properties: $e');
    }
  }

  /// Track a generic event
  static Future<void> track(String eventName, {Map<String, dynamic>? properties}) async {
    if (!_isInitialized || !_analyticsEnabled || _mixpanel == null) return;

    try {
      final eventProps = {
        'timestamp': DateTime.now().toIso8601String(),
        'platform': defaultTargetPlatform.name,
        ...?properties,
      };

      await _mixpanel!.track(eventName, properties: eventProps);
      debugPrint('Analytics event tracked: $eventName');
    } catch (e) {
      debugPrint('Error tracking event $eventName: $e');
    }
  }

  // ========== ONBOARDING EVENTS ==========

  /// Track signup started
  static Future<void> trackSignupStarted(String method) async {
    await track('signup_started', properties: {'method': method});
  }

  /// Track signup completed
  static Future<void> trackSignupCompleted(String userId, String method) async {
    await identify(userId);
    await track('signup_completed', properties: {'method': method});
    await setUserProperties({
      'signup_method': method,
      'signup_date': DateTime.now().toIso8601String(),
    });
  }

  /// Track personality quiz started
  static Future<void> trackPersonalityQuizStarted() async {
    await track('personality_quiz_started');
  }

  /// Track personality quiz completed
  static Future<void> trackPersonalityQuizCompleted() async {
    await track('personality_quiz_completed');
    await setUserProperties({
      'personality_quiz_completed_at': DateTime.now().toIso8601String(),
    });
  }

  /// Track profile setup started
  static Future<void> trackProfileSetupStarted() async {
    await track('profile_setup_started');
  }

  /// Track profile completed
  static Future<void> trackProfileCompleted(String userId) async {
    await identify(userId);
    await track('profile_completed');
    await setUserProperties({
      'profile_completed_at': DateTime.now().toIso8601String(),
      'profile_status': 'complete',
    });
  }

  // ========== MATCHING EVENTS ==========

  /// Track daily selection viewed
  static Future<void> trackDailySelectionViewed(int profileCount) async {
    await track('daily_selection_viewed', properties: {
      'profile_count': profileCount,
    });
  }

  /// Track profile viewed in daily selection
  static Future<void> trackProfileViewed(String profileId) async {
    await track('profile_viewed', properties: {
      'profile_id': profileId,
    });
  }

  /// Track profile chosen (liked)
  static Future<void> trackProfileChosen(String profileId, {double? compatibilityScore}) async {
    await track('profile_chosen', properties: {
      'profile_id': profileId,
      if (compatibilityScore != null) 'compatibility_score': compatibilityScore,
    });
  }

  /// Track profile passed (disliked)
  static Future<void> trackProfilePassed(String profileId) async {
    await track('profile_passed', properties: {
      'profile_id': profileId,
    });
  }

  /// Track match created
  static Future<void> trackMatchCreated(String matchId, String otherUserId) async {
    await track('match_created', properties: {
      'match_id': matchId,
      'other_user_id': otherUserId,
    });
  }

  // ========== CHAT EVENTS ==========

  /// Track chat accepted
  static Future<void> trackChatAccepted(String chatId, String matchId) async {
    await track('chat_accepted', properties: {
      'chat_id': chatId,
      'match_id': matchId,
    });
  }

  /// Track chat declined
  static Future<void> trackChatDeclined(String matchId) async {
    await track('chat_declined', properties: {
      'match_id': matchId,
    });
  }

  /// Track first message sent
  static Future<void> trackFirstMessageSent(String chatId) async {
    await track('first_message_sent', properties: {
      'chat_id': chatId,
    });
  }

  /// Track message sent (any message)
  static Future<void> trackMessageSent(String chatId, {int? messageLength}) async {
    await track('message_sent', properties: {
      'chat_id': chatId,
      if (messageLength != null) 'message_length': messageLength,
    });
  }

  /// Track chat expired
  static Future<void> trackChatExpired(String chatId, int messageCount) async {
    await track('chat_expired', properties: {
      'chat_id': chatId,
      'message_count': messageCount,
    });
  }

  // ========== SUBSCRIPTION EVENTS ==========

  /// Track subscription page viewed
  static Future<void> trackSubscriptionPageViewed() async {
    await track('subscription_page_viewed');
  }

  /// Track subscription started
  static Future<void> trackSubscriptionStarted(String tier, String period) async {
    await track('subscription_started', properties: {
      'tier': tier,
      'period': period,
    });
    await setUserProperties({
      'subscription_tier': tier,
      'subscription_period': period,
      'subscription_started_at': DateTime.now().toIso8601String(),
    });
  }

  /// Track subscription cancelled
  static Future<void> trackSubscriptionCancelled(String tier) async {
    await track('subscription_cancelled', properties: {
      'tier': tier,
    });
    await setUserProperties({
      'subscription_status': 'cancelled',
      'subscription_cancelled_at': DateTime.now().toIso8601String(),
    });
  }

  /// Track subscription restored
  static Future<void> trackSubscriptionRestored(String tier) async {
    await track('subscription_restored', properties: {
      'tier': tier,
    });
  }

  // ========== APP LIFECYCLE EVENTS ==========

  /// Track app opened
  static Future<void> trackAppOpened() async {
    await track('app_opened');
  }

  /// Track app backgrounded
  static Future<void> trackAppBackgrounded() async {
    await track('app_backgrounded');
  }

  // ========== RESET ==========

  /// Reset analytics (for logout/account deletion)
  static Future<void> reset() async {
    if (_mixpanel == null) return;

    try {
      await _mixpanel!.reset();
      debugPrint('Analytics reset');
    } catch (e) {
      debugPrint('Error resetting analytics: $e');
    }
  }
}

File: main-api/src/modules/analytics/analytics.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as Mixpanel from 'mixpanel';
import { GdprService } from '../gdpr/gdpr.service';

export interface AnalyticsEvent {
  name: string;
  userId?: string;
  properties?: Record<string, any>;
  timestamp?: Date;
}

export interface OnboardingEvent {
  step:
    | 'registration'
    | 'profile_creation'
    | 'questionnaire'
    | 'photos_upload'
    | 'completed';
  method?: 'google' | 'apple' | 'email';
  metadata?: Record<string, any>;
}

export interface MatchingEvent {
  action:
    | 'daily_selection_viewed'
    | 'profile_chosen'
    | 'match_created'
    | 'match_expired';
  matchId?: string;
  compatibilityScore?: number;
  metadata?: Record<string, any>;
}

export interface ChatEvent {
  action: 'conversation_started' | 'message_sent' | 'conversation_expired';
  conversationId?: string;
  metadata?: Record<string, any>;
}

export interface SubscriptionEvent {
  action:
    | 'subscription_started'
    | 'subscription_renewed'
    | 'subscription_cancelled'
    | 'subscription_upgraded';
  plan?: string;
  price?: number;
  currency?: string;
  metadata?: Record<string, any>;
}

@Injectable()
export class AnalyticsService {
  private readonly logger = new Logger(AnalyticsService.name);
  private mixpanel: Mixpanel.Mixpanel | null = null;
  private enabled: boolean;

  constructor(
    private readonly configService: ConfigService,
    private readonly gdprService: GdprService,
  ) {
    const token = this.configService.get<string>('analytics.mixpanel.token');
    this.enabled =
      this.configService.get<boolean>('analytics.mixpanel.enabled') || false;

    if (!token || !this.enabled) {
      this.logger.warn('Mixpanel analytics not configured or disabled');
      return;
    }

    try {
      this.mixpanel = Mixpanel.init(token);
      this.logger.log('Mixpanel analytics service initialized');
    } catch (error) {
      const errorMessage = error instanceof Error ? error.stack : String(error);
      this.logger.error('Failed to initialize Mixpanel', errorMessage);
      this.enabled = false;
    }
  }

  /**
   * Check if user has opted out of analytics via GDPR consent
   */
  private async hasAnalyticsConsent(userId: string): Promise<boolean> {
    try {
      const consent = await this.gdprService.getCurrentConsent(userId);

      // If no consent record exists, default to opt-out (GDPR compliant)
      if (!consent) {
        return false;
      }

      // Check if analytics consent is explicitly given
      return consent.analytics === true;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.stack : String(error);
      this.logger.error(
        `Failed to check analytics consent for user ${userId}`,
        errorMessage,
      );
      // On error, default to opt-out for GDPR compliance
      return false;
    }
  }

  /**
   * Track a generic event
   */
  async trackEvent(event: AnalyticsEvent): Promise<void> {
    if (!this.enabled || !this.mixpanel) {
      this.logger.debug('Analytics disabled, skipping event tracking');
      return;
    }

    try {
      // Check GDPR consent if userId is provided
      if (event.userId) {
        const hasConsent = await this.hasAnalyticsConsent(event.userId);
        if (!hasConsent) {
          this.logger.debug(`User ${event.userId} has opted out of analytics`);
          return;
        }
      }

      const properties: Record<string, unknown> = {
        ...event.properties,
        timestamp: event.timestamp || new Date(),
        environment: this.configService.get('app.environment'),
      };

      if (event.userId) {
        this.mixpanel.track(event.name, {
          distinct_id: event.userId,
          ...properties,
        });
      } else {
        this.mixpanel.track(event.name, properties);
      }

      this.logger.debug(`Event tracked: ${event.name}`, {
        userId: event.userId,
      });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.stack : String(error);
      this.logger.error(`Failed to track event: ${event.name}`, errorMessage);
    }
  }

  /**
   * Identify a user in Mixpanel
   */
  async identifyUser(
    userId: string,
    userProperties?: Record<string, unknown>,
  ): Promise<void> {
    if (!this.enabled || !this.mixpanel) {
      return;
    }

    try {
      const hasConsent = await this.hasAnalyticsConsent(userId);
      if (!hasConsent) {
        return;
      }

      this.mixpanel.people.set(userId, {
        $name: userProperties?.name,
        $email: userProperties?.email,
        ...userProperties,
        last_seen: new Date(),
      });

      this.logger.debug(`User identified: ${userId}`);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.stack : String(error);
      this.logger.error(`Failed to identify user: ${userId}`, errorMessage);
    }
  }

  /**
   * Track onboarding events
   */
  async trackOnboarding(userId: string, event: OnboardingEvent): Promise<void> {
    await this.trackEvent({
      name: `onboarding_${event.step}`,
      userId,
      properties: {
        step: event.step,
        method: event.method,
        ...event.metadata,
      },
    });
  }

  /**
   * Track matching events
   */
  async trackMatching(userId: string, event: MatchingEvent): Promise<void> {
    await this.trackEvent({
      name: `matching_${event.action}`,
      userId,
      properties: {
        action: event.action,
        matchId: event.matchId,
        compatibilityScore: event.compatibilityScore,
        ...event.metadata,
      },
    });
  }

  /**
   * Track chat/messaging events
   */
  async trackChat(userId: string, event: ChatEvent): Promise<void> {
    await this.trackEvent({
      name: `chat_${event.action}`,
      userId,
      properties: {
        action: event.action,
        conversationId: event.conversationId,
        ...event.metadata,
      },
    });
  }

  /**
   * Track subscription events
   */
  async trackSubscription(
    userId: string,
    event: SubscriptionEvent,
  ): Promise<void> {
    await this.trackEvent({
      name: `subscription_${event.action}`,
      userId,
      properties: {
        action: event.action,
        plan: event.plan,
        price: event.price,
        currency: event.currency,
        ...event.metadata,
      },
    });

    // Also update user profile with subscription status
    if (
      event.action === 'subscription_started' ||
      event.action === 'subscription_renewed'
    ) {
      await this.identifyUser(userId, {
        subscription_plan: event.plan,
        is_subscriber: true,
      });
    } else if (event.action === 'subscription_cancelled') {
      await this.identifyUser(userId, {
        is_subscriber: false,
      });
    }
  }

  /**
   * Opt user out of analytics (GDPR)
   */
  async optOut(userId: string): Promise<void> {
    if (!this.enabled || !this.mixpanel) {
      return;
    }

    try {
      // Mixpanel opt-out
      this.mixpanel.people.delete_user(userId);
      this.logger.log(`User ${userId} opted out of analytics`);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.stack : String(error);
      this.logger.error(`Failed to opt out user ${userId}`, errorMessage);
    }
  }

  /**
   * Delete all analytics data for a user (GDPR - Right to be Forgotten)
   */
  async deleteUserData(userId: string): Promise<void> {
    if (!this.enabled || !this.mixpanel) {
      return;
    }

    try {
      // Delete user profile and all associated data in Mixpanel
      this.mixpanel.people.delete_user(userId);
      this.logger.log(`Analytics data deleted for user ${userId}`);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.stack : String(error);
      this.logger.error(
        `Failed to delete analytics data for user ${userId}`,
        errorMessage,
      );
    }
  }
}
